backend

DorianSyruss opened the pull request.

Commit:
Add initial config files, basic server and routing

- linters, gitignore, editorconfig, express server, routes

Commit:
Add basic schemas, models, routes

- added nodemon for auto server restart
- added 2 basic routs quotes, users
- get i post methods for quotes

Commit:
Add more API end points, global error handling

- added more api routes
- errs propagating to next middleware to global err handler

Commit:
Improve project structure, implement router

- code restructred to functional units, added api and models folders
- add more routues and api end points
- router implemented

Commit:
Add QuoteCollection model, env config, API/index.js

- app.js refactored

Commit:
Add more  model routes, add more API points, WIP

- added more user routes and api points,
- add option to bookmark quotes and
create quote collections,
- poslih app.js,
- code refactored

Commit:
Add option to list all quotes inside of collection

- promise.all, generators, yield implemented

Commit:
Polish routes, WIP

- new routes added, old routes polished

Commit:
env setup, routes refactor, models refactor WIP

- added more api routes,
- refactor api routes code,
- setup production env with remote db on mongolab

Commit:
facebook login, WIP

Commit:
facebook login/auth WIP, WIP

- code is written in poor manners, for testing purposes
- login passes, but no user is written in db, needs testing and
refactoring

Commit:
Add facebook login, code refactored

Commit:
Remove .env.prod from git

Commit:
Delete .env.prod

Commit:
PR input changes, facebook auth refactored

- code polished
- facebook auth updated, refactored

Commit:
facebook login refactor

Commit:
Implement local strategy login, WIP

- password needs hashing and comparing

Commit:
Login refactor, add password hashing

- WIP, needs polish

Commit:
Implement cookies

-needs testing

Commit:
Implement login, authorization

- WIP

Commit:
Implement auth routes, auth check, basic user roles

- added roles (user, admin), authentication check
- code structure refactored, code polished
- env.template file implemented, fb data now in env, removed credentials
- added db.js fil
- started on seed.js mock-up data

- not all routes are protected (maybe some kind of whitelist instead of middleweare
function in each route:)

Commit:
Protect private routes, seed data for db (WIP)

Commit:
Add seed data WIP

Commit:
Refactor seed.js

vladimyr commented on .editorconfig:
```diff
@@ -0,0 +1,17 @@
+indent_size = 2
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
```
If you do this inside markdown files you'll change the way they get rendered. Always use this paired with:
```ini
[*.md]
trim_trailing_whitespace = false
```

DorianSyruss replied:
didn't know this, thx for the tip 👍

vladimyr commented on .editorconfig:
```diff
@@ -0,0 +1,17 @@
+max_line_length = 120
+
+[*.{js,json}]
+max_line_length = off
```
Specifying max line length through editorconfig setting is bit _controversal_ but if you do that stick to it. What made you turning it off for all code (js) files? :confused:

DorianSyruss replied:
I understand why it's controversial, but my thinking process was that if this project gets "bigger" and some other developers would jump in, maybe it's best for the overall readability to stick to this convection, I don't know. And for the part where I turned it off it was and oversight :)

vladimyr commented on .env.template:
```diff
@@ -0,0 +1,15 @@
+FB_APP_SECRET= #secretKey
+FB_CALLBACK_URL= #url
+
+
```
It is simply matter of style :bowtie: 
Remove excessive blank line.

vladimyr commented on .eslintrc.json:
```diff
@@ -0,0 +1,37 @@
+          "arraysInObjects": false,
+          "objectsInObjects": false,
+          "propertyName": true
+        }],*/
```
:warning: If you use json form of eslint config then you can't use comments inside. Json does not allow comments.

DorianSyruss replied:
You are absolutely right...

vladimyr commented on .eslintrc.json:
```diff
@@ -0,0 +1,37 @@
+    "space-in-parens": ["error", "never"],
+
+
+    // override default options for rules from base configurations
```
:arrow_up: Same as above.

vladimyr commented on .eslintrc.json:
```diff
@@ -0,0 +1,37 @@
+    "comma-dangle": ["error", "never"],
+    "no-cond-assign": ["error", "always"],
+
+    // disable rules from base configurations
```
:arrow_up: Same as above.

vladimyr commented on .gitignore:
```diff
@@ -0,0 +1,7 @@
+dist/
+.idea/
+.env.prod
+*.env
```
What should `*.env` match? :confused:

vladimyr commented on api/auth.js:
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+const passport = require('../auth/index');
```
Well, if you omit `index` and use just `../auth` it will assume `index.js` as require target...

vladimyr commented on api/auth.js:
```diff
@@ -0,0 +1,19 @@
+
+router.get('/auth/facebook', passport.authenticate('facebook', { scope: 'email' }));
+router.get('/auth/facebook/callback', passport.authenticate('facebook',
+  { successRedirect: '/api/users', failureRedirect: '/login' }));
```
Take a step back and think about authentication flow for a second:
* new tab gets opened
* facebook asks you to authorize application to access your data (in this case `email`)
* after you approve/reject authorization request callback url gets called inside current tab

What will happen here is:
a) If user approves authorization request next thing he'll see is JSON list of all user data
b) If user rejects authorization request next thing he'll see is login screen (now in completely new tab)

Both routes are equally bad due to following. Never-ever redirect user to an **api** route. Api routes are for web/mobile clients, server rendered views are for humans. Additionally, keep in mind that you are inside new tab. Whatever happens you should instruct user to close it afterwards. If authentication succeeds present him with nice message and ask him to close it, if it fails tell him it failed and ask him to close it and try again.
i.e. You need one or two dedicated static views to redirect to :wink:

DorianSyruss replied:
Yes I wholeheartedly agree on this issue, the matter here is that I really didn't think this through enough but I knew I needed to implement this in a better way, but I just assumed that I will take care of it later, when I have more knowledge of the matter, or even when the frontend implementation was more clear.

DorianSyruss replied:
And so I did take care of it in [this particular commit](https://github.com/DorianSyruss/wall-of-sayings/pull/1/commits/eeef20751b50776e686c98b725b3475791da42e9), since it happened when I started working on frontend it created a little confusion here, so you can check it out here, or later when you check out later commit. And for the static routes, I will implement that.

vladimyr commented on api/auth.js:
```diff
@@ -0,0 +1,19 @@
+  { successRedirect: '/api/users', failureRedirect: '/login' }));
+
+router.post('/auth/login', passport.authenticate('local', { failureRedirect: '/api/quotes',
+  successRedirect:'/api/users' }));
```
As I already said, never ever redirect to an api route. Yes, you may say that you use it from your rest client only and it saves you from doing additional request but:
1. Sometimes that next request would be targeting something else
2. APIs should be flexible and don't assume anything in general

Posting to `/auth/login` should give you either user profile data or appropriate error.

**PS** It is really awkward to redirect somewhere (besides login page) in case of failure because how would you know if your request succeeded? :dizzy_face: 
And you miss space between colon and `/api/users` :bowtie:

DorianSyruss replied:
same as above 👍

vladimyr commented on api/auth.js:
```diff
@@ -0,0 +1,19 @@
+router.get('/auth/logout', (req, res) => {
+  req.session.destroy(() => {
+    res.redirect('/api/quotes');
+  });
```
No need for block here, you can inline it. Less lines is better :wink:

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+    .catch(err => next(err));
+}
+
+function createQuoteCollection(req, res, next){
```
Space :tired_face:

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+}
+
+function createQuoteCollection(req, res, next){
+  let collection = Object.assign({}, req.body, { owner: req.query.owner });
```
So you are sending all data needed to create new collection through body (using post) except of owner id that comes as part of url by means of query param?

![](https://media.giphy.com/media/3o7TKF5TmcHvHjCVlC/giphy.gif)

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+    .catch(err => next(err));
+}
+
+function deleteQuoteCollection(req, res, next){
```
Space :tired_face: :angry:

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+}
+
+function addNewQuote(req, res, next) {
+    QuoteCollection.findById(req.params.id)
```
Indentation is off :straight_ruler: :red_circle:

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+}
+
+function deleteQuote(req, res, next) {
+  let quote_ids = req.body.quote_ids || [];
```
This sort of safety guard can be moved to `#deleteQuote` implementation by adding default argument value.

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+}
+
+function addNewCollaborator(req, res, next) {
+  let collaborator_ids = req.body.collaborator_ids || [];
```
:arrow_up: Same as above.

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+}
+
+function removeCollaborators(req, res, next){
+  let collaborator_ids = req.body.collaborator_ids || [];
```
:arrow_up: Same as above.

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+function listCollectionQuotes(req, res, next) {
+  QuoteCollection.findById(req.params.id)
+    .then(collection => collection.getQuotes())
+    .then(quotes => { res.status(HTTPStatus.OK).send(quotes); })
```
Using block here? Wat?

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+  let collaborator_ids = req.body.collaborator_ids || [];
+  QuoteCollection.findById(req.params.id)
+    .then(quoteCollection => quoteCollection.addCollaborators(collaborator_ids))
+    .then(queryStatus => res.status(HTTPStatus.OK).send(queryStatus))
```
`queryStatus` means literally nothing :sob: Be more descriptive, I don't want to enter `#addCollaborators` implementation in order to figure out what it returns.

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+function addNewQuote(req, res, next) {
+    QuoteCollection.findById(req.params.id)
+      .then(quoteCollection => quoteCollection.addQuote(req.body.quote_id))
+      .then(queryStatus => res.status(HTTPStatus.OK).send(queryStatus))
```
Missed this `queryStatus` during first pass, check later comment...

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+  let collaborator_ids = req.body.collaborator_ids || [];
+  QuoteCollection.findById(req.params.id)
+    .then(quoteCollection => quoteCollection.removeCollaborators(collaborator_ids))
+    .then(queryStatus => res.status(HTTPStatus.OK).send(queryStatus))
```
`queryStatus` strikes again:
![](https://media.giphy.com/media/134WAWHsoUpZV6/giphy.gif)

vladimyr commented on api/collections.js:
```diff
@@ -0,0 +1,92 @@
+    .catch(err => next(err));
+}
+
+
```
Every time you put one or more extra blank lines:
**A puppy dies!** :dog:

vladimyr commented on api/index.js:
```diff
@@ -0,0 +1,45 @@
+}, {
+  path: '/api/quotes',
+  method: 'GET'
+}];
```
According to this setup all routes except `/api/collections` ones are public. If you have narrower protected space invert your security approach?

vladimyr commented on api/index.js:
```diff
@@ -0,0 +1,45 @@
+    return;
+  }
+  isLoggedIn(req, res, next);
+}
```
This is actually useless? because of following reasons:
1. Inside `collections` subrouter you already have global `isLoggedIn` middleware applied and that is actually only non-public route collection
2. Checking if user is logged isn't very secure way to protect stuff. You are protecting users from general public but not from each other?

vladimyr commented on api/index.js:
```diff
@@ -0,0 +1,45 @@
+  app.use('/api', users);
+  app.use('/api', quotes);
+  app.use('/api', quoteCollections);
+};
```
What you have is :bow_and_arrow: and what you need is :rocket: 
Instead of reinventing wheel try to utilize something like: https://github.com/ForbesLindesay/connect-roles

vladimyr commented on api/quotes.js:
```diff
@@ -0,0 +1,45 @@
+    .catch(err => next(err));
+}
+
+function createQuote(req, res, next){
```
Space :angry: :rage:

vladimyr commented on api/quotes.js:
```diff
@@ -0,0 +1,45 @@
+
+function createQuote(req, res, next){
+  let { quote, author } = req.body;
+  Quote.create({ quote, author })
```
Why destructure it anyway?

DorianSyruss replied:
What about security aspect of this, body can contain garbage, and Quote.create saves it to base directly??

vladimyr commented on api/quotes.js:
```diff
@@ -0,0 +1,45 @@
+
+function updateQuote(req, res, next) {
+  let update = { quote: req.body.quote, author: req.body.author };
+  Quote.findOneAndUpdate(req.params.id, update, { new: true })
```
Just send `req.body` without destructuring...

vladimyr commented on api/users.js:
```diff
@@ -0,0 +1,45 @@
+const User = require('../models/user');
+const { isAdmin } = require('../auth/permissions');
+
+router.get('/users', isAdmin, listUsers);
```
Typically all logged users can see who is around but not in the same level of detail as admin does :wink:

vladimyr commented on api/users.js:
```diff
@@ -0,0 +1,45 @@
+router.get('/users', isAdmin, listUsers);
+router.post('/register', createUser);
+router.get('/users/:id', getUser);
+router.put('/users/:id', updateUser);
```
:warning: Anyone logged in can update any user if he is able to guess his id...
Authorization needs to be revised.

vladimyr commented on api/users.js:
```diff
@@ -0,0 +1,45 @@
+const { isAdmin } = require('../auth/permissions');
+
+router.get('/users', isAdmin, listUsers);
+router.post('/register', createUser);
```
These are just my 2c but this actually sounds more like: `/auth/signup`.

vladimyr commented on api/users.js:
```diff
@@ -0,0 +1,45 @@
+
+function updateUser(req, res, next) {
+  let update = { name: req.body.name, surname: req.body.surname };
+  User.findOneAndUpdate(req.params.id, update, { new: true })
```
Do not destructure, simply pass body in...

vladimyr commented on app.js:
```diff
@@ -0,0 +1,41 @@
+app.use(bodyParser.urlencoded({ extended: true }));
+
+app.use(session({
+  secret: 'keyboard cat',
```
This is serious security issue :exclamation: 
This *must* be hidden inside `.env` file. And please use bit stronger secret...

vladimyr commented on app.js:
```diff
@@ -0,0 +1,41 @@
+app.use(passport.initialize());
+app.use(passport.session());
+
+app.use(express.static(__dirname + '/public'));
```
Never ever concatenate paths with plus :angry:

vladimyr commented on app.js:
```diff
@@ -0,0 +1,41 @@
+  res.status(HTTPStatus.INTERNAL_SERVER_ERROR).json({ error: err.message });
+});
+
+//set port
```
Space, please :cry:

vladimyr commented on app.js:
```diff
@@ -0,0 +1,41 @@
+
+// tell the app to listen on specified port
+app.listen(port);
+console.log('Server running on port: ' + port);
```
Use template string :wink:

vladimyr commented on auth/facebook.js:
```diff
@@ -0,0 +1,32 @@
+
+    User.create({
+      facebookId: profile.id,
+      email: profile.emails[0].value,
```
I don't think returned profile contains both `facebook_id` and `id` :grey_exclamation: 
Also, you used lodash/get and then completely annihilated it by accessing it all over again in highly unsafe way :exclamation:

DorianSyruss replied:
😯  true true...

vladimyr commented on auth/index.js:
```diff
@@ -0,0 +1,18 @@
+passport.use(local);
+passport.use(facebook);
+passport.serializeUser((user, done) => {
+  done(null, user.id);
```
Remove curlies...

vladimyr commented on auth/index.js:
```diff
@@ -0,0 +1,18 @@
+  done(null, user.id);
+});
+passport.deserializeUser((id, done) => {
+  User.findById(id, done);
```
:arrow_up: Same as above.

vladimyr commented on auth/permissions.js:
```diff
@@ -0,0 +1,28 @@
+const HTTPStatus = require('http-status');
```
![](https://i.imgflip.com/1q9mkw.jpg)

vladimyr commented on package.json:
```diff
@@ -0,0 +1,35 @@
+    "bcrypt": "^1.0.2",
+    "bluebird": "^3.5.0",
+    "dotenv": "^4.0.0",
+    "ejs": "^2.5.6",
```
Not used :dash:

vladimyr commented on package.json:
```diff
@@ -0,0 +1,35 @@
+    "http-status": "^1.0.1",
+    "lodash": "^4.17.4",
+    "mongoose": "^4.9.2",
+    "nib": "^1.1.2",
```
Not used :dash:

vladimyr commented on package.json:
```diff
@@ -0,0 +1,35 @@
+  "author": "DorianSyruss",
+  "license": "ISC",
+  "devDependencies": {
+    "body-parser": "^1.17.1",
```
Certainly not a dev one...

vladimyr reviewed (changes requested):
It could be better but not bad at all.
Areas of improvement:
1) Style and (stop) relying on linter to warn you about breaking style conventions :angry: 
2) Naming stuff (sometimes it is way to generic) :no_mouth: 
3) Not thinking thoroughly while implementing functionality (in this case authorization) :zipper_mouth_face:

Commit:
Code refactor according to the input, WIP

Commit:
Code refactor according to PR input

Commit:
Delete .DS_Store

Commit:
Implement authorization, wip

Commit:
Refactor roles authorization, WIP

- implemented Connect Roles module,
- added roles auth, code refactored, needs more testing

Commit:
Merge branch 'backend' of github.com:DorianSyruss/wall-of-sayings into backend

Commit:
Implemented auth, refactored routes, WIP

- some err handlers need more testing

Commit:
Add custom operational error msg

-implement operational error to add quote
to collection route,
- did more testing

Commit:
Refactor quote favorite counter logic

- implement quote favorites counter in the way
that it tracks users that favorited the quotes or deleted it
instead of simply tracking the 'addToCollection' event

Commit:
Implement role check, implement basic login test

- started writing tests
- implemented role check bug on getQuoteCollection route

Commit:
Refactor REST api routes (make them atomic)

- refactor routes on api/collections
- code polish / remove extensive code

Commit:
Refactor REST api routes, code polish

- refactor routes on api/users
- omit private data from db
- resolve path to public folder

Commit:
Further refactor REST api logic for users.js

Commit:
Refactor REST api routes for quotes, WIP

- separate admin, guest and user route logic
- create quote types, update seeds

Commit:
Refactor api routes for quote collections, optimize route logic

- get /me routes for personal user data
- optimize error checking in route handlers
- permissions changed, WIP (waiting to refactor user and quote routes)
- implement getCollaborators method for listing all collaborating users

Commit:
Refactor user api routes

- get /me routes for personal user data
- optimize error checking in route handlers
- change permissions WIP

Commit:
Implement password hashing middleware

- refactor pass hashing

Commit:
Refactor quotes api routes, refactor permission logic

- refactor api routes for quotes
- implement more security checks
- remove excessive code for permission logic

Commit:
Implement native mongoose data tracking, quote tags

- implemented createdAt, updatedAt data tracking
- implement tracking when quote went public

Commit:
Refator seed.js script for password hashing

- implement new way of hashing the seed data user
password because 'save' hook was removed when the middleware
for hashing was introduced

Commit:
Fix facebook login bug

- fix and refactor finding the facebookId in db

Commit:
Implement server pagination, add missing routes, start validation

- implement offset and limit for big data chunks
- implement missing routes (for public collections)
- start implementing data validation
- implement validation error checking on rest api data

Commit:
Refactor password hashing, implement password validation

- implement user password validation strategy
- refactor hasing, remove hashing middleware

Commit:
Implement further validation on models WIP

Commit:
Add models validation, implement more core API features, add package-lock.json

- improve general models validation
- add quote collections validation
- implement email validator
- remove owner when quote is going public
- filter out quotes from collections that were deleted from DB by Admin
- refactor reset password feature

vladimyr commented on api/collections.js:
```diff
@@ -211,7 +211,7 @@ function removeQuoteFromMyCollection(req, res, next) {
       return quoteCollection.removeQuote(req.body.quoteId, req.user.id)
         .then(status => res.status(HTTPStatus.OK).send(status))
-        .error((e) => res.status(HTTPStatus.BAD_REQUEST).send(e.message));
+        .error(e => res.status(HTTPStatus.BAD_REQUEST).send(e.message));
```
Let's bring in some _standard_ conventions here:

- `e` means event
- `err` means error

Also you might sometimes find `reason` as `catch` parameter name because Promise could technically be rejected by non-error value bad that's code smell...

vladimyr commented on api/users.js:
```diff
@@ -108,6 +108,7 @@ function resetPassword(req, res, next) {
       return user.updatePassword(req.body.oldPassword, req.body.password);
     })
     .then(() => res.status(HTTPStatus.NO_CONTENT).send())
+    .error(e => res.status(HTTPStatus.BAD_REQUEST).send(e.message))
```
:arrow_up: Look up, same applies here...

vladimyr commented on models/quote.js:
```diff
@@ -41,7 +51,9 @@ Object.assign(quote.methods, {
-    this.publishedAt = new Date;
+    if (delete this.owner) {
+      this.publishedAt = new Date;
+    }
```
What should `#trackPublishing` do? It seems poorly named? :confused: 
Is it supposed to publish something publicly? Why not `#publish` or `#makePublic`?

Besides I'm having hard time understanding that `delete` snippet. According to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete) `delete` operator returns:
>true for all cases except when the property is an own non-configurable property, in which case, false is returned in non-strict mode.

Which pretty much means in your case it is absolutely always true :dizzy_face: 
Putting that aside for a moment, deleting stuff is simply wrong way of doing thing. It is one of slowest V8 operations and you don't need it here. Instead simply put its value to `undefined`.

vladimyr commented on models/quoteCollection.js:
```diff
@@ -7,10 +7,15 @@ const { ObjectId } = Schema;
-  title: String,
-  description: String,
+  title: { type: String, required: true, maxlength: 140 },
+  description: { type: String, maxlength: 500 },
```
Be consistent about formatting your schemas. I would stick with multiline everywhere... :bowtie:

vladimyr commented on models/quote.js:
```diff
@@ -41,7 +51,9 @@ Object.assign(quote.methods, {
 
   trackPublishing(type) {
     if (type !== 'public') return;
```
Magic string:
![](https://media.giphy.com/media/l4FGIgsVPdoRd2wbS/200w.gif)

Use some sort of map or enum :wink:

vladimyr commented on models/quoteCollection.js:
```diff
@@ -47,7 +49,8 @@ Object.assign(quoteCollection.methods, {
     const Quote = mongoose.model('Quote');
-    return Promise.all(this.quotes.map(id => Quote.findById(id)));
+    return Promise.all(this.quotes.map(id => Quote.findById(id)))
+      .then(quotes => quotes.filter((quote) => quote !== null));
```
This is suboptimal from database perspective. You are looping over array of ids asking mongo each time to retrieve single document. Instead you should use `Quote.find` with array of possible ids (issues single mongo query). :warning: 
Also, this filter looks weird :dizzy_face:

vladimyr commented on models/user.js:
```diff
@@ -2,7 +2,9 @@
 
 const bcrypt = require('bcrypt');
 const mongoose = require('mongoose');
+const validator = require('validator');
```
You have to be kiddin' me :upside_down_face: 
Bringing in library that has among all other stuff freakin' postal code validation in order to check email? You just killed a puppy :dog: :gun:

vladimyr commented on models/user.js:
```diff
@@ -21,13 +23,25 @@ const Role = {
+    required: true,
+    minlength: 2,
+    maxlength: 100
+  },
```
Man who made stackoverflow has really interesting blog and among all other niceties he assembled list of new programming jargon: https://blog.codinghorror.com/new-programming-jargon/
Note **3. Egyptian Brackets**. It is used to describe following indent scheme:
```c
if (a == b) {
  printf("hello");
}
```
However you just invented _new direction_: **Egyptian objects**
```js
name: { type: String,
  required: true,
  minlength: 2,
  maxlength: 100
},
```

vladimyr reviewed (changes requested):
I might take long pauses but don't worry :wink:

![](https://media.giphy.com/media/LqW9dLVjQm3cs/200w.gif)

DorianSyruss commented:
Thank you for your time mr. @vladimyr , I was already aware of some of these issues, I'll do my best to resolve them and to implement few more core features and improvements so we can move on to frontend 👍

Commit:
Refactor models, api routes, validation according to PR input, WIP

- remove magic string appearances like 'private', 'public' ...
- improve standard naming convetions (e, err, ...)
- improve code style formating in schemas
- refactor method for publishing quotes (make it more clear of it's purpose)
- make necessary changes in route handlers to accommodate model changes
- make validation run on every model update

vladimyr commented on api/collections.js:
```diff
@@ -135,7 +135,7 @@ function updateMyQuoteCollection(req, res, next) {
   const update = dropProperties(req.body, immutables);
   const query = { owner: req.user.id, _id: req.params.id };
-  QuoteCollection.findOneAndUpdate(query, update, { new: true })
+  QuoteCollection.findOneAndUpdate(query, update, { new: true, runValidators: true })
```
Well, if you extracted everything into local vars do that for options too :bowtie:

vladimyr commented on api/collections.js:
```diff
@@ -254,7 +254,7 @@ function getQuoteCollection(req, res, next) {
 function updateQuoteCollection(req, res, next) {
   const data = dropProperties(req.body, immutables);
-  QuoteCollection.findOneAndUpdate(req.params.id, data, { new: true })
+  QuoteCollection.findOneAndUpdate(req.params.id, data, { new: true, runValidators: true })
```
Same as above ⬆️

vladimyr commented on api/quotes.js:
```diff
@@ -4,6 +4,7 @@ const Quote = require('../models/quote');
 const router = require('express').Router();
 const HTTPStatus = require('http-status');
 const dropProperties = require('lodash/omit');
+const Types = require('../models/quote').types;
```
How about:
```js
const { Types } = require('../models/quote');
```

vladimyr commented on models/quoteCollection.js:
```diff
@@ -4,17 +4,25 @@ const mongoose = require('mongoose');
     required: true,
-    enum: ['private', 'public'],
-    default: 'private'
+    enum: [Types.Private, Types.Public],
```
How about using `lodash#toArray`
```js
enum: toArray(Types)
```

Commit:
Code polish and refactor according to PR input

- switch email validator
- refactor Types from method to static object
- polish code in collections api

Commit:
Code polish

- code polish in users and quotes api

vladimyr commented on api/collections.js:
```diff
@@ -286,7 +288,7 @@ function removeQuoteFromCollection(req, res, next) {
       return quoteCollection.removeQuote(req.body.quoteId, req.user.id)
         .then(status => res.status(HTTPStatus.OK).send(status))
-        .error((e) => res.status(HTTPStatus.BAD_REQUEST).send(e.message));
+        .error(e => res.status(HTTPStatus.BAD_REQUEST).send(e.message));
```
Let me write it again: `err` for errors, `e` for events :wink:

vladimyr commented on api/quotes.js:
```diff
@@ -4,7 +4,7 @@ const Quote = require('../models/quote');
 const HTTPStatus = require('http-status');
 const dropProperties = require('lodash/omit');
-const Types = require('../models/quote').types;
+const { Types } = require('../models/quote');
```
:+1:

vladimyr commented on models/user.js:
```diff
@@ -2,7 +2,7 @@
 const bcrypt = require('bcrypt');
 const mongoose = require('mongoose');
-const validator = require('validator');
+const emailValidator = require('email-validator');
```
Ok, this is way better than previous solution, keep it as is but just for sake of argument next time take author background into account. For example, I personally tend to have more faith in:

- hapijs folks - https://github.com/hapijs/isemail
- [segment.io](https://segment.com/) - https://github.com/segmentio/is-email

DorianSyruss replied:
Valid point, I'll keep this in mind.

Commit:
Refactor quoteColections methods, code polish

- refactor method for listing quotes and collaborators from quote collections
- code polish

vladimyr commented on models/quoteCollection.js:
```diff
@@ -75,7 +74,7 @@ Object.assign(quoteCollection.methods, {
     const privateData = ['password'];
     const User = mongoose.model('User');
-    return Promise.all(this.collaborators.map(id => User.findById(id).omit(privateData)));
+    return User.find({ _id: { $in: this.collaborators } }).omit(privateData);
```
It would be nice to create some sort of helper:
```js
Model.findManyById(arrayOfIds)
```
so you can write stuff like:
```js
Quote.findManyById(this.quotes);
```
and:
```js
User.findManyById(this.collaborators)...
```

Commit:
Add more helper methods, refactor collection api handlers, code polish

- add findManyById helper method on mongoose model
- remove use of 'public' string across the code base, use Types
- make only public quotes available in public collections

vladimyr commented on api/collections.js:
```diff
@@ -74,20 +75,26 @@ function getPublicCollection(req, res, next) {
+           else return quote.type;
+         }))
+          .then(publicQuotes => res.status(HTTPStatus.OK).send(publicQuotes));
+        });
```
Something is not right here. What are you trying to achieve with this block of code? 😕

DorianSyruss replied:
Uh, since you have to ask, I guess it's no good :) Anyway let me try to explain. This route fetches and returns all quotes from the public quote collection. What I mean by public quote collections is a collection that user decided to be available to general public audience. But also we have to block quotes that are private, because only public quotes (what I mean by public is approved by admin through voting process) must be seen and rendered. This could be solved by not allowing the collection to go public if it contains any private quotes, but i prefer this way instead. No reason to block the whole collections to go public because of a few unapproved (private) quotes. By returning quote.type back when the quote is private, we have knowledge that more quotes exist in collection but they are not available to public yet.

vladimyr replied:
So all this can be _translated_ as `collection.getQuotes({ type: Types.Public })` :wink:

DorianSyruss replied:
ahahah that solution crossed my mind long before this whole "shabang" but you as you can remember I modified the getQuotes so it calls another method findManyById which has to return all the quotes because it's not used only for this purpose, but anyway I will try to implement one more query parameter to pass along or something... `  getQuotes() {
    const Quote = mongoose.model('Quote');
    return Quote.findManyById(this.quotes);
  },` 
Such a nice example how one optimization makes another more complex .... 😞

vladimyr commented on models/helpers.js:
```diff
@@ -30,3 +31,10 @@ passwordSchema
+  findManyById(arrayOfIds) {
+    return this.find({ _id: { $in: arrayOfIds } });
+  }
+});
```
Nice 👍

Commit:
Refactor api handlers and methods, add filtering options

- refactor method for getting all public quotes from public collection
- enable filtering on collections and quotes

Commit:
Refactor method for getting multiple objects by id on model

vladimyr commented on models/helpers.js:
```diff
@@ -33,13 +33,9 @@ passwordSchema
-    return this.find({ _id: { $in: arrayOfIds }, type: Types.Public });
+  getById(arrayOfIds, type) {
+    const conditions = { _id: { $in: arrayOfIds }, type };
+    return this.find(omitBy(conditions, prop => !prop));
```
Create generic method:
```js
findMany(query = {}) { /*...*/ }
```
And handle query properties in following way:

- If property `value` is array use `prop: { $in: value }` form, otherwise use simply `prop: value`
- If property `value` is '*' remove it from `query` (using omitBy)

Omitting stuff by checking for _falsey_ value works in this case but in general it is bad idea because asking for `undefined` and/or `false` value inside query is legit.

Commit:
Add findMany helper method

- refactor getQuotes and getCollaborators methods on quoteCollections

Commit:
Add filltering by collaborators in collections

- refactor getMyCollections route, add new filter
- add isCollaboratedBy method for later permissions implementation

Commit:
Refactor rest API logic for quoteCollections

- refactor permissions and getQuoteCollection route

Commit:
Refactor quoteCollections rest api

- implement new routing logic

Commit:
Change api collections structure

- break into smaller modules

Commit:
Refactor quoteCollection api routes

- switch more routes to use setContext method

Commit:
Refactor quoteCollections api routes

- add setContextMany function for listing routes
- refactor listing routes according to changes in logic
- refactor permissions, add can('list') permission that uses setContextMany

Commit:
Remove excessive code

- remove public.js under api/collections

Commit:
Remove excessive code for quoteCollections api

- delete user.js file under api/collections

Commit:
Refactor quotes api, apply new routes logic

- refactor to setContext logic
- implement setContextOne and setContextMany
- use setContext type functions as helpers to reuse them for other routes
- remove excessive code after refactoring
